extensions/opentelemetry/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/grpc/GrpcTracingServerInterceptor.java

```diff
@@ -0,0 +1,180 @@
+
+        GrpcRequest grpcRequest = GrpcRequest.server(call.getMethodDescriptor(), headers, call.getAttributes());
+        Context parentContext = Context.current();
+        Context spanContext = null;
+        Scope scope = null;
+        boolean shouldStart = instrumenter.shouldStart(parentContext, grpcRequest);
+        if (shouldStart) {
+            spanContext = instrumenter.start(parentContext, grpcRequest);
+            scope = spanContext.makeCurrent();
+        }
+
+        try {
+            TracingServerCall<ReqT, RespT> tracingServerCall = new TracingServerCall<>(call, spanContext, grpcRequest);
+            return new TracingServerCallListener<>(next.startCall(tracingServerCall, headers), spanContext, grpcRequest);
+        } finally {
+            if (scope != null) {
+                scope.close();
+            }
+        }
+    }
+
+    private static class GrpcServerNetServerAttributesGetter extends InetSocketAddressNetServerAttributesGetter<GrpcRequest> {
+        @Override
+        public InetSocketAddress getAddress(final GrpcRequest grpcRequest) {
+            SocketAddress socketAddress = grpcRequest.getAttributes().get(Grpc.TRANSPORT_ATTR_REMOTE_ADDR);
+            if (socketAddress instanceof InetSocketAddress) {
+                return (InetSocketAddress) socketAddress;
+            }
+            return null;
+        }
+
+        @Override
+        public String transport(final GrpcRequest grpcRequest) {
+            return SemanticAttributes.NetTransportValues.IP_TCP;
+        }
+    }
+
+    private static class GrpcTextMapGetter implements TextMapGetter<GrpcRequest> {
+        @Override
+        public Iterable<String> keys(final GrpcRequest carrier) {
+            return carrier.getMetadata() != null ? carrier.getMetadata().keys() : Collections.emptySet();
+        }
+
+        @Override
+        public String get(final GrpcRequest carrier, final String key) {
+            if (carrier != null && carrier.getMetadata() != null) {
+                return carrier.getMetadata().get(Metadata.Key.of(key, Metadata.ASCII_STRING_MARSHALLER));
+            } else {
+                return null;
+            }
+        }
+    }
+
+    private class TracingServerCallListener<ReqT> extends SimpleForwardingServerCallListener<ReqT> {
+        private final Context spanContext;
+        private final GrpcRequest grpcRequest;
+
+        protected TracingServerCallListener(final ServerCall.Listener<ReqT> delegate, final Context spanContext,
+                final GrpcRequest grpcRequest) {
+            super(delegate);
+            this.spanContext = spanContext;
+            this.grpcRequest = grpcRequest;
+        }
+
+        @Override
+        public void onHalfClose() {
+            try {
+                super.onHalfClose();
+            } catch (Exception e) {
+                instrumenter.end(spanContext, grpcRequest, null, e);
```

Comment:
I believe you may need to de-duplicate the call to "end".

Reply:
From what I've seen, `ServerCall#close` will always be called either directly by `onHalfClose` or later by `onComplete`.